Kernel/FileSystem/VFSRootContext.cpp

```diff
@@ -116,32 +116,223 @@ ErrorOr<NonnullRefPtr<VFSRootContext>> VFSRootContext::create_with_empty_ramfs()
+            });
+            return {};
+        });
+    });
+}
+
+ErrorOr<void> VFSRootContext::do_full_teardown(Badge<PowerStateSwitchTask>)
+{
+    // NOTE: We are going to tear down the entire VFS root context from its mounts.
+    // To do this properly, we swap out the original root custody with the empty
+    // root custody for vfs root context of kernel processes.
+    m_root_custody.with([](auto& custody) {
+        custody = VFSRootContext::empty_context_custody_for_kernel_processes();
+    });
+
+    auto unmount_was_successful = true;
+    while (unmount_was_successful) {
+        unmount_was_successful = false;
+        Vector<Mount&, 16> mounts;
+        TRY(m_details.with([&mounts](auto& details) -> ErrorOr<void> {
+            for (auto& mount : details.mounts) {
+                TRY(mounts.try_append(const_cast<Mount&>(mount)));
+            }
+            return {};
+        }));
+        if (mounts.is_empty())
+            break;
+        auto const remaining_mounts = mounts.size();
+
+        while (!mounts.is_empty()) {
+            auto& mount = mounts.take_last();
+            TRY(mount.guest_fs().flush_writes());
+
+            auto mount_path = TRY(mount.absolute_path());
+            auto& mount_inode = mount.guest();
+            auto const result = VirtualFileSystem::unmount(*this, mount_inode, mount_path->view());
+            if (result.is_error()) {
+                dbgln("Error during unmount of {}: {}", mount_path, result.error());
+                // FIXME: For unknown reasons the root FS stays busy even after everything else has shut down and was unmounted.
+                //        Until we find the underlying issue, allow an unclean shutdown here.
+                if (remaining_mounts <= 1)
+                    dbgln("BUG! One mount remaining; the root file system may not be unmountable at all. Shutting down anyways.");
+            } else {
+                unmount_was_successful = true;
+            }
+        }
+    }
+    return {};
+}
+
+ErrorOr<void> VFSRootContext::unmount(FileBackedFileSystem::List& file_backed_file_systems_list, Inode& guest_inode, StringView custody_path)
+{
+    return m_details.with([&](auto& details) -> ErrorOr<void> {
+        bool did_unmount = false;
+        for (auto& mount : details.mounts) {
+            if (&mount.guest() != &guest_inode)
+                continue;
+            auto mountpoint_path = TRY(mount.absolute_path());
+            if (custody_path != mountpoint_path->view())
+                continue;
+
+            TRY(VFSRootContext::validate_mount_not_immutable_while_being_used(details, mount));
+            dbgln("VFSRootContext({}): Unmounting {}...", id(), custody_path);
+            TRY(VirtualFileSystem::remove_mount(mount, file_backed_file_systems_list));
+            did_unmount = true;
+            break;
+        }
+        if (!did_unmount) {
+            dbgln("VirtualFileSystem: Nothing mounted on inode {}", guest_inode.identifier());
```

Comment:
nit?: Shouldn't the message be updated to say `VFSRootContext` now? (same for the other one in this function)

Reply:
Right. I will fix this in the next revision.